### PR TITLE
Extra customization options.

### PR DESCRIPTION
**Pull request by JohnBobbo96**
- I added an option to be able to sign kernel modules during the build using [Arch-SKM](https://github.com/itoffshore/Arch-SKM).
- I have also added an option to be able to change what nice level to compile at with a limit of -8 being highest and 8 being lowest, and also the ionice level from 0 to 7.
- I have added an option to enable [CCACHE_FILECLONE](https://ccache.dev/manual/latest.html#config_file_clone) which allows ccache to attempt to take advantage of 'CoW', and 'reflinks', etc, and also [CCACHE_INODECACHE](https://ccache.dev/manual/latest.html#config_inode_cache) which when enabled lets ccache store source file hashes based on device, inode and timestamps. And can possibly reduce compilation time.
- I also made it so that instead of deleting log files they instead get moved to a file with a '-old' extension.

I have tested it with Archlinux on both PDS and BMQ
```
$ uname -r
5.19.11-269-tkg-pds

$ zgrep MODULE /proc/config.gz
CONFIG_MODULE_SIG_FORMAT=y
CONFIG_MODULES=y
CONFIG_MODULE_FORCE_LOAD=y
CONFIG_MODULE_UNLOAD=y
CONFIG_MODULE_FORCE_UNLOAD=y
CONFIG_MODULE_UNLOAD_TAINT_TRACKING=y
CONFIG_MODULE_SIG=y
CONFIG_MODULE_SIG_FORCE=y
CONFIG_MODULE_SIG_ALL=y
# CONFIG_MODULE_SIG_SHA1 is not set
# CONFIG_MODULE_SIG_SHA224 is not set
# CONFIG_MODULE_SIG_SHA256 is not set
# CONFIG_MODULE_SIG_SHA384 is not set
CONFIG_MODULE_SIG_SHA512=y
CONFIG_MODULE_SIG_HASH="sha512"
# CONFIG_MODULE_COMPRESS_NONE is not set
# CONFIG_MODULE_COMPRESS_GZIP is not set
# CONFIG_MODULE_COMPRESS_XZ is not set
CONFIG_MODULE_COMPRESS_ZSTD=y
# CONFIG_MODULE_ALLOW_MISSING_NAMESPACE_IMPORTS is not set
CONFIG_MODULES_TREE_LOOKUP=y
CONFIG_MODULE_SIG_KEY="certs/signing_key.pem"
CONFIG_MODULE_SIG_KEY_TYPE_RSA=y
# CONFIG_MODULE_SIG_KEY_TYPE_ECDSA is not set
CONFIG_DEBUG_INFO_BTF_MODULES=y
# CONFIG_MODULE_ALLOW_BTF_MISMATCH is not set
```
```
$ uname -r
5.19.11-269-tkg-bmq

$ zgrep MODULE /proc/config.gz
CONFIG_MODULE_SIG_FORMAT=y
CONFIG_MODULES=y
CONFIG_MODULE_FORCE_LOAD=y
CONFIG_MODULE_UNLOAD=y
CONFIG_MODULE_FORCE_UNLOAD=y
CONFIG_MODULE_UNLOAD_TAINT_TRACKING=y
CONFIG_MODULE_SIG=y
CONFIG_MODULE_SIG_FORCE=y
CONFIG_MODULE_SIG_ALL=y
# CONFIG_MODULE_SIG_SHA1 is not set
# CONFIG_MODULE_SIG_SHA224 is not set
# CONFIG_MODULE_SIG_SHA256 is not set
# CONFIG_MODULE_SIG_SHA384 is not set
CONFIG_MODULE_SIG_SHA512=y
CONFIG_MODULE_SIG_HASH="sha512"
# CONFIG_MODULE_COMPRESS_NONE is not set
# CONFIG_MODULE_COMPRESS_GZIP is not set
# CONFIG_MODULE_COMPRESS_XZ is not set
CONFIG_MODULE_COMPRESS_ZSTD=y
# CONFIG_MODULE_ALLOW_MISSING_NAMESPACE_IMPORTS is not set
CONFIG_MODULES_TREE_LOOKUP=y
CONFIG_MODULE_SIG_KEY="certs/signing_key.pem"
CONFIG_MODULE_SIG_KEY_TYPE_RSA=y
# CONFIG_MODULE_SIG_KEY_TYPE_ECDSA is not set
CONFIG_DEBUG_INFO_BTF_MODULES=y
# CONFIG_MODULE_ALLOW_BTF_MISMATCH is not set
```

and these are there the logs from each build.
[linux-john-pds.log](https://github.com/Frogging-Family/linux-tkg/files/9647245/linux-john-pds.log) [linux-john-bmq.log](https://github.com/Frogging-Family/linux-tkg/files/9647247/linux-john-bmq.log)

### CODE REVIEW

File: linux-tkg-config/prepare [hunk 1]
```diff
@@ -462,18 +462,38 @@ _linux_git_branch_checkout() {
...
-    msg2 "please delete the current linux-tkg folder and re-created it with"
-    msg2 "git clone --depth=1 https://github.com/Frogging-Family/linux-tkg.git"
+  if ! git status &>/dev/null; then
+    warning "Please make sure you use git to clone the repository."
+    msg2 "Try deleting '$PWD' and cloning the repo with git."
+    msg2 "Example: 'git clone --depth=1 https://github.com/Frogging-Family/linux-tkg.git'"
     exit 1
   fi
 
+  if [[ -x "$(command -v schedtool)" ]]; then
```
Comment (JohnBobbo96) [inline linux-tkg-config/prepare:477]: This should hopefully fix #580.

**Comment by JohnBobbo96** [comment]
- There is now an option to download a release tarball from kernel.org if they do not wish to use a git clone for any reason, and also an option to verify signatures/checksums of the tarball or git tag.
- The user's strip flags from makepkg.conf will now be used to also strip modules.
- The community-patches repository is now cloned using depth 1.
- I added the same ascii frog from PKGBUILD to install.sh
- Install.sh now respects the users choice of _use_tmpfs and _tmpfs_path.
- There is now a sha256sums-tkg.asc file with the sha256sums of all of the files inside linux-tkg-config and linux-tkg-patches to verify sources in makepkg and also install.sh
- The user can now specify how many threads to use for compilation by turning _force_all_threads to false and adjusting _thread_num.
- Linus Torvalds github repository is now an option for _git_mirror.

I have tested all these changes with,
PDS - Using git and makepkg ([linux-pds-git-makepkg.log](https://github.com/Frogging-Family/linux-tkg/files/9691812/linux-pds-git-makepkg.log)),
BMQ - Using release tarball and makepkg ([linux-bmq-makepkg.log](https://github.com/Frogging-Family/linux-tkg/files/9691815/linux-bmq-makepkg.log)),
BORE - Using git and install.sh ([linux-bore-git-install.sh.log](https://github.com/Frogging-Family/linux-tkg/files/9691817/linux-bore-git-install.sh.log)),
BORE - Using release tarball and install.sh ([linux-bore-install.sh.log](https://github.com/Frogging-Family/linux-tkg/files/9691819/linux-bore-install.sh.log.diff)).

On a side-note i was wondering if you guys were aware of [Kernel.org Finger Banner](https://kernel.org/finger_banner) which could be used to get the latest kernel version easier but i personally don't know how to implement it.

Any feedback would be appreciated.

File: linux-tkg-config/prepare [hunk 2]
```diff
@@ -462,18 +462,38 @@ _linux_git_branch_checkout() {
...
+    fi
+    if [[ "$_ionice_level" -gt "7" ]]; then
+      _ionice_level="7"
+    elif [[ "$_ionice_level" -lt "0" ]]; then
+      _ionice_level="0"
+    elif [[ -z "$_ionice_level" ]]; then
+      _ionice_level="1"
+    fi
+  else
+    _use_schedtool="false"
```
[the remark this answers is not in the document]
Comment (Tk-Glitch) [reply linux-tkg-config/prepare:641]: It should

**Comment by AdelKS** [comment]
Yeah it was about time to have the Frog logo shown in `install.sh` :D

I haven't tried your `ccache` changes, I assume they improve the hit rate ?

About bringing backs tars and verifying their signature: I do not see at all why tars would be useful, except for CIs that need to clone freshly the kernel sources, which will make them take few extra minutes on a compile that will take most of the time anyway. For regular users that keep their linux-tkg clone around it's better from any standpoint.

About signing modules, it appears to me as a rather very specific use-case, it's for people who do not use luks2 but still want to guarantee the integrity of the kernel (needs TPM2 otherwise it's useless since the kernel image itself can be compromised/changed before booting into it). I had a quick look and it looks like the signing approach you want to bring up can work with official arch kernels too ? That means it can come _after_ having generated the kernel's packages ?

Mr Torvald's Github repository does not contain the stable branches, which means it does not contain all the necessary tags to fetch specific non-master kernel versions (like e.g. `5.10.72`). So we cannot add that.

**Pull request closed without merging**